gastime/gastime.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (C) 2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// Package gastime measures time based on the consumption of gas.
+package gastime
+
+import (
+	"github.com/ava-labs/avalanchego/vms/components/gas"
+	"github.com/holiman/uint256"
+
+	"github.com/ava-labs/strevm/intmath"
+	"github.com/ava-labs/strevm/proxytime"
+)
+
+// Time represents an instant in time, its passage measured in [gas.Gas]
+// consumption. It is not thread safe nor is the zero value valid.
+//
+// In addition to the passage of time, it also tracks excess consumption above a
+// target, as described in [ACP-194] as a "continuous" version of [ACP-176].
+//
+// Copying a Time, either directly or by dereferencing a pointer, will result in
+// undefined behaviour. Use [Time.Clone] instead as it reestablishes internal
+// invariants.
+//
+// [ACP-176]: https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/176-dynamic-evm-gas-limit-and-price-discovery-updates
+// [ACP-194]: https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/194-streaming-asynchronous-execution
+type Time struct {
+	TimeMarshaler
+}
+
+// makeTime is a constructor shared by [New] and [Time.Clone].
+func makeTime(t *proxytime.Time[gas.Gas], target, excess gas.Gas) *Time {
+	tm := &Time{
+		TimeMarshaler: TimeMarshaler{
+			Time:   t,
+			target: target,
+			excess: excess,
+		},
+	}
+	tm.establishInvariants()
+	return tm
+}
+
+func (tm *Time) establishInvariants() {
+	tm.Time.SetRateInvariants(&tm.target, &tm.excess)
+}
+
+// New returns a new [Time], set from a Unix timestamp. The consumption of
+// `2*target` units of [gas.Gas] is equivalent to a tick of 1 second.
+func New(unixSeconds uint64, target, startingExcess gas.Gas) *Time {
+	return makeTime(proxytime.New(unixSeconds, 2*target), target, startingExcess)
+}
+
+// Clone returns a deep copy of the time.
+func (tm *Time) Clone() *Time {
+	// [proxytime.Time.Clone] explicitly does NOT clone the rate invariants, so
+	// we reestablish them as if we were constructing a new instance.
+	return makeTime(tm.Time.Clone(), tm.target, tm.excess)
+}
+
+// Target returns the `T` parameter of ACP-176.
+func (tm *Time) Target() gas.Gas {
+	return tm.target
+}
+
+// Excess returns the `x` variable of ACP-176.
+func (tm *Time) Excess() gas.Gas {
+	return tm.excess
+}
+
+// Price returns the price of a unit of gas, i.e. the "base fee".
+func (tm *Time) Price() gas.Price {
+	return gas.CalculatePrice(1 /* M */, tm.excess, 87*tm.target /* K */)
```

Comment:
When calculating the conversion constant, we have an unchecked multiplication by 87 which is currently capped: https://github.com/ava-labs/coreth/blob/master/plugin/evm/upgrade/acp176/acp176.go#L93